src/lib/transition_router/initial_validator.ml

```diff
@@ -233,142 +232,109 @@ module Duplicate_block_detector = struct
           [%log error] ~metadata msg )
 end
 
-let run ~logger ~trust_system ~verifier ~transition_reader
-    ~valid_transition_writer ~initialization_finish_signal ~precomputed_values =
-  let genesis_state_hash =
-    (Precomputed_values.genesis_state_hashes precomputed_values).state_hash
-  in
-  let genesis_constants =
-    Precomputed_values.genesis_constants precomputed_values
-  in
-  let rejected_blocks_logger =
-    Logger.create ~id:Logger.Logger_id.rejected_blocks ()
-  in
-  let open Deferred.Let_syntax in
-  let duplicate_checker = Duplicate_block_detector.create () in
-  O1trace.background_thread "initially_validate_blocks" (fun () ->
-      Reader.iter transition_reader
-        ~f:(fun
-             ( `Transition transition_env
-             , `Time_received time_received
-             , `Valid_cb valid_cb )
-           ->
-          let state_hash =
-            ( Envelope.Incoming.data transition_env
-            |> Mina_block.header |> Header.protocol_state
-            |> Protocol_state.hashes )
-              .state_hash
+let validate ~logger ~trust_system ~verifier ~initialization_finish_signal
+    ~precomputed_values =
+  stage (fun ~transition_env ~time_received ~valid_cb ->
+      let genesis_state_hash =
+        (Precomputed_values.genesis_state_hashes precomputed_values).state_hash
+      in
+      let genesis_constants =
+        Precomputed_values.genesis_constants precomputed_values
+      in
+      let rejected_blocks_logger =
+        Logger.create ~id:Logger.Logger_id.rejected_blocks ()
+      in
+      let open Deferred.Let_syntax in
+      let duplicate_checker = Duplicate_block_detector.create () in
```

Comment:
This doesn't seem to be preserved between calls, right?

Reply:
Fixed.